Client/trunk/ParaEngineClient/NPL/NPLConnection.cpp

```diff
@@ -702,13 +710,73 @@ void NPL::CNPLConnection::handleDisconnect( int reason )
 	snprintf(msg_reason, 255, "%d", reason);
 	m_msg_dispatcher.PostNetworkEvent(NPL_ConnectionDisconnected, GetNID().c_str(), msg_reason);
 }
+bool NPL::CNPLConnection::handle_websocket_data(int bytes_transferred)
+{
+	WebSocket::ByteBuffer b = WebSocket::WebSocketReader::load(&m_buffer, bytes_transferred);
+	if (m_websocket_reader.parse(b))
+	{
+		NPL::WebSocket::WebSocketFrame* frame = m_websocket_reader.getFrame();
+
+		vector<byte> data = frame->getData();
+		
+
+		NPL::WebSocket::OpCode opcode = (NPL::WebSocket::OpCode)frame->getOpCode();
+		switch (opcode)
+		{
+		case NPL::WebSocket::TEXT:
+		{
+			string code(data.begin(), data.end());
+
+			Json::Value root;
+			Json::Reader reader;
+			Json::FastWriter writer;
+			bool parsingSuccessful = reader.parse(code, root);     //parse process
+			if (!parsingSuccessful)
+			{
+				return false;
+			}
+			int server_id = root["s_id"].asInt();
+			Json::Value msg = root["msg"];
+			string msg_str = writer.write(msg);
+			m_input_msg.method = "A";
+			m_input_msg.m_n_filename = server_id;
+
+			string m_code;
+			NPL::NPLHelper::EncodeJsonStringInQuotation(m_code,0, msg_str);
+			m_input_msg.m_code = m_code;
+			handleMessageIn();
+			break;
+		}
+		case NPL::WebSocket::BINARY:
+			break;
+		case NPL::WebSocket::CLOSE:
+			stop();
+			break;
+		case NPL::WebSocket::PING:
+			break;
+		case NPL::WebSocket::PONG:
+			break;
+		default:
+			break;
+		}
+		return true;
+	}
+	return false;
+}
 
 bool NPL::CNPLConnection::handleReceivedData( int bytes_transferred )
 {
+	// first try to parse websocket protocol
+	if (handle_websocket_data(bytes_transferred))
```

Comment:
this is NOT right. You should remember the current protocol state in a variable in NPLConnection, instead of checking the buffer data to determine the type.  When protocol is upgraded, we need to remember it.